File: answers/leetcode/Peeking Iterator.cpp
//@type Design
//@result 12 / 12 test cases passed. Status: Accepted Runtime: 4 ms Submitted: 1 minute ago

// Below is the interface for Iterator, which is already defined for you.
// **DO NOT** modify the interface for Iterator.
class Iterator {
    struct Data;
	Data* data;
public:
	Iterator(const vector<int>& nums);
	Iterator(const Iterator& iter);
	virtual ~Iterator();
	// Returns the next element in the iteration.
	int next();
	// Returns true if the iteration has more elements.
	bool hasNext() const;
};


class PeekingIterator : public Iterator {
public:
	PeekingIterator(const vector<int>& nums) : Iterator(nums) {
	    // Initialize any member here.
	    // **DO NOT** save a copy of nums and manipulate it directly.
	    // You should only use the Iterator interface methods.
	    head = 0;
		peeked = false;
	}

    // Returns the next element in the iteration without advancing the iterator.
	int peek() {
        if (! peeked) {
			head = next();
			peeked = true;
		}
		return head;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	int next() {
        if (peeked) {
			peeked = false;
			return head;
		}
		else {
			return Iterator::next();
		}
	}

	bool hasNext() const {
        if (peeked) {
			return true;
		}
		else {
			return Iterator::hasNext();
		}
	}
	int head;
	bool peeked;
};

File: answers/leetcode/Peeking Iterator/Peeking Iterator.java
// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html
class PeekingIterator implements Iterator<Integer> {
    
    private Integer nextInt;
    private Iterator<Integer> iterator;

	public PeekingIterator(Iterator<Integer> iterator) {
	    // initialize any member here.
	    this.iterator = iterator;
        this.nextInt = null;
        this.next();
	}

    // Returns the next element in the iteration without advancing the iterator.
	public Integer peek() {
        return this.nextInt;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	@Override
	public Integer next() {
	    Integer result = this.nextInt;
        this.nextInt = this.iterator.hasNext() ? this.iterator.next() : null;
        return result;
	}

	@Override
	public boolean hasNext() {
	    return this.nextInt != null;
	}
}